src/dispatcher.rs

```diff
@@ -0,0 +1,162 @@
+use std::sync::Arc;
+
+#[cfg(feature = "prefetch")]
+use ethcore::transaction::Action;
+use ethcore::transaction::SignedTransaction;
+use failure::{Error, Fail, Fallible, ResultExt};
+use serde_bytes::ByteBuf;
+
+use oasis_core_keymanager_client::KeyManagerClient;
+#[cfg(feature = "prefetch")]
+use oasis_core_runtime::storage::mkvs::Prefix;
+use oasis_core_runtime::{
+    common::{cbor, crypto::hash::Hash, roothash::Message as RoothashMessage},
+    transaction::{
+        context::Context,
+        dispatcher::{CheckOnlySuccess, Dispatcher as TxnDispatcher},
+        tags::Tags,
+        types::*,
+    },
+};
+
+use super::{
+    block::OasisBatchHandler,
+    methods::{check, execute},
+};
+
+use oasis_runtime_api as api;
+
+/// Dispatch error.
+#[derive(Debug, Fail)]
+enum DispatchError {
+    #[fail(display = "method not found: {}", method)]
+    MethodNotFound { method: String },
+}
+
+pub struct DecodedCall {
+    pub transaction: SignedTransaction,
+}
+
+pub struct Dispatcher {
+    /// Registered batch handler.
+    batch_handler: Box<OasisBatchHandler>,
```

Comment:
```suggestion
    batch_handler: Box<dyn OasisBatchHandler>,
```

or, alternatively
```rust
pub struct Dispatcher<B> {
	batch_handler: Box<B>
}

impl<B: OasisBatchHandler> for Dispatcher<B> {
}
```
if you'd prefer `Dispatcher` to be monomorphized

Reply:
I got rid of the `Box` altogether, since it's not a trait and the `Dispatcher` itself is already boxed.